### PR TITLE
Split dashboards into separate configmaps

### LINKED ISSUE #505: In some k8s distros there is an error thrown when creating dashboards
## Bug Report

**Describe the bug**
In some environments, the generated dashboard configmap may cause issues when being updated or created with `kubectl apply`.

**Expected behavior**
A warning is not thrown and the dashboards are created

**Additional context**
Note this may be resolved with `kubectl create` and `kubectl replace`, but we should do our part to reduce the size of this configmap.

### PR DESCRIPTION
**Pull request by bradfordcp**
**What this PR does**: Splits the Grafana dashboards configmap into separate files

**Which issue(s) this PR fixes**:
Fixes #505

**Checklist**
- [x] Changes manually tested
- [x] Chart versions updated (if necessary)
- [x] Automated Tests added/updated
- [x] Documentation added/updated
- [x] CHANGELOG.md updated (not required for documentation PRs)
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

**Comment by bradfordcp** [comment]
From my local test:
image

With the following available via kubectl

```console
~/s/k8ssandra (enhancements/split-dashboards|✔) $ kubectl get configmaps
NAME                                                          DATA   AGE
k8ssandra-grafana                                             2      2m30s
k8ssandra-k8ssandra-grafana-stargate-dashboard                1      2m30s
k8ssandra-kube-prometheus-grafana-datasource                  1      2m30s
k8ssandra-grafana-config-dashboards                           1      2m30s
k8ssandra-grafana-test                                        1      2m30s
k8ssandra-k8ssandra-grafana-cassandra-condensed-dashboard     1      2m30s
k8ssandra-k8ssandra-grafana-cassandra-overview-dashboard      1      2m30s
k8ssandra-k8ssandra-grafana-system-metrics-dashboard          1      2m30s
cass-operator-lock                                            0      117s
prometheus-k8ssandra-kube-prometheus-prometheus-rulefiles-0   18     27s
~/s/k8ssandra (enhancements/split-dashboards|✔) $ kubectl get configmap k8ssandra-k8ssandra-grafana-stargate-dashboard
NAME                                             DATA   AGE
k8ssandra-k8ssandra-grafana-stargate-dashboard   1      2m34s
~/s/k8ssandra (enhancements/split-dashboards|✔) $ kubectl get configmap k8ssandra-k8ssandra-grafana-cassandra-condensed-dashboard
NAME                                                        DATA   AGE
k8ssandra-k8ssandra-grafana-cassandra-condensed-dashboard   1      2m43s
~/s/k8ssandra (enhancements/split-dashboards|✔) $ kubectl get configmap k8ssandra-k8ssandra-grafana-cassandra-overview-dashboard
NAME                                                       DATA   AGE
k8ssandra-k8ssandra-grafana-cassandra-overview-dashboard   1      2m48s
~/s/k8ssandra (enhancements/split-dashboards|✔) $ kubectl get configmap k8ssandra-k8ssandra-grafana-system-metrics-dashboard
NAME                                                   DATA   AGE
k8ssandra-k8ssandra-grafana-system-metrics-dashboard   1      2m55s
~/s/k8ssandra (enhancements/split-dashboards|✔) $ 
```

File: charts/k8ssandra/templates/grafana/cassandra-condensed.configmap.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{{- if .Values.monitoring.grafana.provision_dashboards }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}-k8ssandra-grafana-cassandra-condensed-dashboard
```
Comment (jsanda) [inline charts/k8ssandra/templates/grafana/cassandra-condensed.configmap.yaml:5]: What do you think about shortening the name to `{{ .Release.Name }}-cassandra-condensed`?

File: charts/k8ssandra/templates/grafana/cassandra-overview.configmap.yaml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{{- if .Values.monitoring.grafana.provision_dashboards }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}-k8ssandra-grafana-cassandra-overview-dashboard
```
Comment (jsanda) [inline charts/k8ssandra/templates/grafana/cassandra-overview.configmap.yaml:5]: What do you think about shortening the name to `{{ .Release.Name }}-cassandra-overview`?

File: charts/k8ssandra/templates/grafana/stargate.configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{{- if (and .Values.monitoring.grafana.provision_dashboards .Values.stargate.enabled) }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}-k8ssandra-grafana-stargate-dashboard
```
Comment (jsanda) [inline charts/k8ssandra/templates/grafana/stargate.configmap.yaml:5]: What do you think about shortening the name to `{{ .Release.Name }}-cassandra-overview`?

File: charts/k8ssandra/templates/grafana/system-metrics.configmap.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+{{- if .Values.monitoring.grafana.provision_dashboards }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}-k8ssandra-grafana-system-metrics-dashboard
```
Comment (jsanda) [inline charts/k8ssandra/templates/grafana/system-metrics.configmap.yaml:5]: What do you think about shortening the name to `{{ .Release.Name }}-cassandra-overview`?

**Review by jakerobb — APPROVED** [review]
LGTM! I'd consider moving the configmap templates into the directories that correspond to the components they monitor, e.g. `templates/grafana/stargate.configmap.yaml` could move to `templates/stargate/grafana-dashboard.yaml`.

**Review by jsanda — APPROVED** [review]
I left a few questions about changing the ConfigMap names, but I am fine merging as is if you want to keep the current names.

Thanks for the fix!

File: charts/k8ssandra/templates/grafana/cassandra-condensed.configmap.yaml (hunk 1 above)
Comment (bradfordcp) [reply charts/k8ssandra/templates/grafana/cassandra-condensed.configmap.yaml:5]: I'd want to keep some sort of indicator that it's for grafana and / or a dashboard.

WDYT?

1. `{{ .Release.Name }}-k8ssandra-grafana-cassandra-condensed-dashboard`
2. `{{ .Release.Name }}-cassandra-condensed`
3. `{{ .Release.Name }}-grafana-cassandra-condensed`
4. `{{ .Release.Name }}-dashboard-cassandra-condensed`
5. `{{ .Release.Name }}-grafana-dashboard-cassandra-condensed`

Comment (bradfordcp) [reply charts/k8ssandra/templates/grafana/cassandra-condensed.configmap.yaml:5]: Note I tried to go from least to most specific, that may be foolish, but it's a convention :shrug:

File: charts/k8ssandra/templates/grafana/system-metrics.configmap.yaml (hunk 4 above)
Comment (bradfordcp) [reply charts/k8ssandra/templates/grafana/system-metrics.configmap.yaml:5]: > LGTM! I'd consider moving the configmap templates into the directories that correspond to the components they monitor, e.g. `templates/grafana/stargate.configmap.yaml` could move to `templates/stargate/grafana-dashboard.yaml`.

This is a good suggestion. @jsanda what are your thoughts on getting rid of the grafana folder altogether?

**Comment by bradfordcp** [comment]
> LGTM! I'd consider moving the configmap templates into the directories that correspond to the components they monitor, e.g. `templates/grafana/stargate.configmap.yaml` could move to `templates/stargate/grafana-dashboard.yaml`.

This is a good suggestion. @jsanda what are your thoughts on getting rid of the grafana folder altogether?

File: charts/k8ssandra/templates/grafana/cassandra-condensed.configmap.yaml (hunk 1 above)
Comment (jsanda) [reply charts/k8ssandra/templates/grafana/cassandra-condensed.configmap.yaml:5]: Thanks for laying out the options. I guess I'd go with 5.

**Comment by jsanda** [comment]
> > LGTM! I'd consider moving the configmap templates into the directories that correspond to the components they monitor, e.g. `templates/grafana/stargate.configmap.yaml` could move to `templates/stargate/grafana-dashboard.yaml`.
> 
> This is a good suggestion. @jsanda what are your thoughts on getting rid of the grafana folder altogether?

👍

**Pull request merged**